activesupport/lib/active_support/error_reporter.rb

```diff
@@ -57,7 +58,11 @@ def handle(error_class = StandardError, severity: :warning, context: {}, fallbac
       yield
     rescue error_class => error
       report(error, handled: true, severity: severity, context: context)
-      fallback
+      if fallback.respond_to?(:call)
+        fallback.call
+      else
+        fallback
+      end
```

Comment:
My worry with these arguments that takes either a callable or not is that you might legitimately want to return an object that respond to call and get bit.

So maybe it should always take a callable?

Reply:
Yeah, to me this feels more likely to give a surprising result than a helpful one.

If we want to accomodate a callable fallback, in this context I'd suggest we consciously choose to check for a `Proc` over `#call`.

At some point it's going to make more sense to just write your invocation as:

```ruby
begin
  user = Rails.error.report do
    # ..
  end
rescue
  user = User.anonymous
end
```

.. or similar (_assuming_ you're actually in a context where `nil ||` is insufficient).

I see the value in providing a neat shortcut for the simplest cases, where we can save a lot of noise, but I'd rather not get too close to inventing a Result API here. I don't even expect this to be a common user-facing API -- mostly something used internally and by framework-supporting libraries -- so I'd also prefer to see in-framework callsites that can benefit, before we overbuild an API ahead of how we're going to use it.

Reply:
Thanks for the feedback. I don't disagree with any of this. I do however disagree with the original change I made regarding ambiguity of "maybe nil is the result, maybe nil is because of a swallowed error" as an API design. I want to avoid that.

So, I see and suggest three possible ways forward, but am open to others:

1) Always require a proc, as in:
```ruby
rescue error_class => error
  report(error, handled: true, severity: severity, context: context)
  fallback.call if fallback
end
```

2) Explicitly check for `Proc` class, as in:
```ruby
rescue error_class => error
  report(error, handled: true, severity: severity, context: context)
  if fallback.is_a?(Proc)
    fallback.call
  else
    fallback
  end
end
```

3) Scrap the `fallback:` idea, and explicitly always return `nil`. This sidesteps the whole issue and prevents users from depending on an ambiguous API:
```ruby
def handle(error_class = StandardError, severity: :warning, context: {}, fallback: nil)
  yield
  nil
rescue error_class => error
  report(error, handled: true, severity: severity, context: context)
  nil
end
```

I know this isn't design by committee consensus, but I don't have much skin in the game here. Though I'm happy to have more and happy to help. @byroot ?

Reply:
I think my preference goes for 1, with maybe an eager raise if you pass a non-callable, so that you don't get bit only once you hit the error path.

But it's not a strongly held opinion. I agree however that the original API introduced some ambiguity.

Reply:
FWIW I find the fallback concept confusing. I am struggling to understand where it should be used.

The examples in https://github.com/rails/rails/pull/43716 all seem like code that shouldn’t raise but should return nil in exceptional cases (eg if no record is found). So I worry we are encouraging use of exceptions for control flow where it’s not really necessary.

Do you have some more examples you can show that the concept was extracted from?

Reply:
@ghiculescu it's usually used for resiliency purpose. Let's imagine you implement a e-commerce website (how original), and you store the shopping cart of users in Redis.

So you have a code like this:

```ruby
class Cart
  def self.fetch(cart_token)
    load(redis.get("cart:#{token}"))
  end
end
```

And then on your storefront you call `Cart.fetch(token)`.

Now if for some reason Redis is down, your entire storefront is down as well. But ultimately your could just fallback to no longer display the cart, and still show the product page.

So it makes something like this:

```ruby
class Cart
  def self.fetch(cart_token)
    Rails.error.handle(Redis::BaseConnectionError, fallback: -> { NullCart.new })
      load(redis.get("cart:#{token}"))
    end
  end
end
```

So it's not exceptions for control flow, even more so because we're reporting the error to a service so it's really not something you'd want to use for the happy path, only for degraded paths.